In [2]:
from __future__ import division
import numpy as np
import os, sys
import matplotlib
#matplotlib.use('Agg')
%matplotlib nbagg
import matplotlib.pyplot as plt
import pandas as pd
import subprocess
from collections import defaultdict, Counter
import pickle

def getCDF(data):
    xdata = np.sort(data)
    ydata = [i/len(xdata) for i in range(len(xdata))]
    return xdata, ydata

## Load consolidated data
### Cleanup
- remove those with missing query type entries
- remove entries with missing timing field
- remove those with invalid query types

In [102]:
df = pd.read_csv('../data/tshark_20170207_20170212_consolidated/split_hour_00000_20170206235900.csv',
                 names=['timestamp', 'time_relative', 'srcIP', 'dstIP', 'qname', 'qtype', 'macAddr', 'subnet', 'host_type', 
'host_name', 'system_type', 'OS', 'nic_manufacturer'])

print len(df)

# remove those with NaN in query type
df = df.dropna(subset=['qtype'])

1936754


In [103]:
# 244 entries contain invalid qtype and need to be cleaned up
# 144 of these seem to be from the same virtual machine source with IP 140.180.219.15 and have 2 dns query separated by a comma
# the rest seem to be querying '<Unknown extended label>' or have missing timestamp fileds and are VMs
df[ ~df['qtype'].str.contains('0x') ]

timestamp   time_relative            srcIP  \
4666              1.48644e+09          7.1883   140.180.219.15   
4960              1.48644e+09         7.59091   140.180.219.15   
5261              1.48644e+09         8.17301   140.180.219.15   
5272              1.48644e+09         8.18228     10.8.147.162   
11366             1.48644e+09         19.2461   140.180.219.15   
11560             1.48644e+09         19.5395   140.180.219.15   
56948             1.48644e+09         87.0393   140.180.219.15   
57143             1.48644e+09         87.3367   140.180.219.15   
57369             1.48644e+09         87.7473   140.180.219.15   
64447             1.48644e+09         99.5771     10.8.162.118   
67786             1.48644e+09         103.876     77.234.44.45   
68373             1.48644e+09         104.536     77.234.44.45   
68951             1.48644e+09         105.505     77.234.44.45   
69305             1.48644e+09         106.056     77.234.44.45   
70187             1.48644e+09         107.348     77.234.44.45   
70324             1.48644e+09         107.596     77.234.44.45   
70340             1.48644e+09         107.639     77.234.44.45   
71262             1.48644e+09         108.973     77.234.44.45   
74764             1.48644e+09         114.415     77.234.44.45   
76074             1.48644e+09         116.636     77.234.44.45   
76357             1.48644e+09         117.081     77.234.44.45   
76440             1.48644e+09         117.188     77.234.44.45   
76441             1.48644e+09         117.188     77.234.44.45   
76751             1.48644e+09         117.726     77.234.44.45   
77034             1.48644e+09         118.188     77.234.44.45   
77035             1.48644e+09         118.189     77.234.44.45   
77036             1.48644e+09         118.189     77.234.44.45   
95288             1.48644e+09         143.633     77.234.44.45   
102630            1.48644e+09         153.987     77.234.44.45   
108710            1.48644e+09         162.331   140.180.219.15   
...                       ...             ...              ...   
1657363           1.48645e+09         3013.88   140.180.219.15   
1657649           1.48645e+09         3014.61   140.180.219.15   
1657926           1.48645e+09         3015.33   140.180.219.15   
1666921           1.48645e+09         3036.42   140.180.219.15   
1670713           1.48645e+09         3044.59   140.180.219.15   
1671627           1.48645e+09         3046.82   140.180.219.15   
1671931           1.48645e+09         3047.54   140.180.219.15   
1681793           1.48645e+09         3067.05   140.180.219.15   
1685294           1.48645e+09         3075.19   140.180.219.15   
1686336           1.48645e+09         3077.63   140.180.219.15   
1686487           1.48645e+09         3077.86   140.180.219.15   
1706329           1.48645e+09         3123.62   140.180.219.15   
1708377           1.48645e+09         3127.31   140.180.219.15   
1708576           1.48645e+09         3127.67   140.180.219.15   
1726166           1.48645e+09         3167.16      10.9.128.80   
1726194           1.48645e+09          3167.2      10.9.128.80   
1728517           1.48645e+09         3171.66     10.8.147.162   
1767710           1.48645e+09         3237.82   140.180.219.15   
1858173  1486446968.944222000  3428.058712000   140.180.219.15   
1858436  1486446969.389055000  3428.503545000   140.180.219.15   
1858708  1486446969.854997000  3428.969487000   140.180.219.15   
1892147                     �      0x0000000c  0:50:56:93:6:6d   
1892149                     �      0x0000000c  0:50:56:93:6:6d   
1892151                     �      0x0000000c  0:50:56:93:6:6d   
1892153                     �      0x0000000c  0:50:56:93:6:6d   
1892159                     �      0x0000000c  0:50:56:93:6:6d   
1930345           1.48645e+09         3586.67   140.180.219.15   
1930487           1.48645e+09         3587.04   140.180.219.15   
1930697           1.48645e+09         3587.39   140.180.219.

In [104]:
# clean up data based on valid query types
df = df[ df['qtype'].str.contains('0x') ]

In [105]:
print "total valid queries =", len(df)

total valid queries = 1935015


### dtype conversions

In [106]:
# convert timestamps to float
df[['timestamp', 'time_relative']] = df[['timestamp', 'time_relative']].apply(pd.to_numeric, errors='coerce')

# convert qtype (A, CNAME) from hex to int
df['qtype'] = df['qtype'].apply(lambda x: int( x.split(',')[0], 16))

# lower case all dns queries
df['qname'] = df['qname'].str.lower()


In [110]:
dnstype_count = df.groupby('qtype')['qname'].count()
dnstype_count.sort_values(inplace=True, ascending=False)

print "Percentage of A Records in data = ", dnstype_count.iloc[0]/dnstype_count.sum()*100

dnstype_count.head(15)

Percentage of A Records in data =  77.8253915344


qtype
1        1505933
28        196059
12        153732
15         37187
33         33905
16          3572
6           3331
43           500
255          319
249          185
2            110
46879         36
41503         14
41247         11
14270          6
Name: qname, dtype: int64

## 1. Filter to A records only

In [112]:
df_a = (df[df.qtype == 1]).copy()

## Add 1LD and 2LD as column

In [115]:
def reduce_domain(domain, level):
    tmp = domain.split('.')
    out = '.'.join( tmp[-level:] )
    return out
    
print reduce_domain('www.youtube.com', 2)
print reduce_domain('www.youtube.com', 1)

youtube.com
com


In [118]:
df_a['1LD'] = df_a['qname'].apply(lambda x: reduce_domain(x, 1))
df_a['2LD'] = df_a['qname'].apply(lambda x: reduce_domain(x, 2))

## 2. Filter out PRINCETON.EDU

In [120]:
df_b = df_a[ df_a['2LD'] != 'princeton.edu' ]
print len(df_a), len(df_b)

1505933 1357654


## Extract srcIP unique features (to act as confirmation of Ground Truth)
- group by source IP as key
- Check if ['macAddr', 'subnet', 'host_type', 'host_name', 'system_type', 'OS', 'nic_manufacturer'] remain the same for each source IP after grouping

In [127]:
check_uniqueness = df_b.groupby('srcIP')[['macAddr', 'subnet', 'host_type', 'host_name', 'system_type',
                                          'OS', 'nic_manufacturer']].aggregate(lambda x: len(set(x)))

check_uniqueness[check_uniqueness['macAddr'] > 1]

Empty DataFrame
Columns: [macAddr, subnet, host_type, host_name, system_type, OS, nic_manufacturer]
Index: []

In [131]:
srcIP_groundtruth = df_b.groupby('srcIP')[['macAddr', 'subnet', 'host_type', 'host_name', 'system_type',
                                          'OS', 'nic_manufacturer']].first()
srcIP_groundtruth.head()

macAddr          subnet host_type   host_name  \
srcIP                                                                  
10.6.48.10   b8:ca:3a:6e:ec:58  fw-servernet-2      host       win04   
10.6.48.100     0:50:56:93:8:e  fw-servernet-2      host   dev-webqa   
10.6.48.105   0:50:56:93:7f:61  fw-servernet-2      host   sdpvm200w   
10.6.48.112    0:50:56:93:6a:6  fw-servernet-2      host  spcacs201w   
10.6.48.116   0:50:56:93:68:af  fw-servernet-2      host  socopm100w   

                 system_type               OS nic_manufacturer  
srcIP                                                           
10.6.48.10           dell-pc  ms-windows-2012             Dell  
10.6.48.100  virtual-machine  ms-windows-2012           Vmware  
10.6.48.105  virtual-machine  ms-windows-2012           Vmware  
10.6.48.112  virtual-machine  ms-windows-2012           Vmware  
10.6.48.116  virtual-machine  ms-windows-2008           Vmware

## Extract Features from rest of data with key=srcIP
- only use relevant colums 'timestamp', 'time_relative', 'srcIP', 'dstIP', 'qname', '1LD', '2LD'

In [137]:
df_c = df_b[['timestamp', 'time_relative', 'srcIP', 'dstIP', 'qname', '1LD', '2LD']]
grouped = df_c.groupby('srcIP')

In [145]:
# grouped.agg({'C' : np.sum, 'D' : lambda x: np.std(x, ddof=1)})
data = defaultdict(list)

for srcip, group in grouped:
    # key
    data['srcIP'].append( srcip )
    
    # DSTIP
    # number of unique dstip
    data['dstIP.count'].append( len( set( group['dstIP'] ) ) )
    
    # most common dstip
    data['dstIP.most_popular'].append( Counter( group['dstIP'] ).most_common(1)[0][0] )
    
    # DNS QRY
    # count of total domains
    data['qname.count'].append( group['qname'].count() )
    
    # count of unique domains
    data['qname.unique.count'].append( len( set( group['qname'] )) )
    
    # count of unique 1LDs
    data['1LD.unique.count'].append( len( set( group['1LD'] )) )
    # count of unique 2LDs
    data['2LD.unique.count'].append( len( set( group['2LD'] )) )
    
    # TIME DIFF ALL - DESCRIBE
    td = group['timestamp'].diff()  #fold(1)
    data['time_diff.mean'].append( td.mean() )
    data['time_diff.std'].append( td.std() )
    data['time_diff.min'].append( td.min() ) 
    data['time_diff.max'].append( td.max() )
    data['time_diff.median'].append( td.median() )
    
    # TOP 3 most popular domains and their time diffs
    top3 = Counter( group['qname'] ).most_common(3)
    for num in range(3):
        column_name = 'qname.most_pop'+ str(num)
        
        if len(top3)>num:
            pop_domains = top3[num]
            domain = pop_domains[0]  #domain
            occurance = pop_domains[1]   #number of lookups
            data[column_name + '.domain'].append( domain )   
            data[column_name + '.count'].append( occurance )
            # time diff
            td = group[ group['qname']==domain ]['timestamp'].diff()
            data[column_name + '.time_diff.mean'].append( td.mean() )
            data[column_name + '.time_diff.std'].append( td.std() )
            data[column_name + '.time_diff.min'].append( td.min() )
            data[column_name + '.time_diff.max'].append( td.max() )
            data[column_name + '.time_diff.median'].append( td.median() )
            
        else:
            data[column_name + '.domain'].append( np.nan )   
            data[column_name + '.count'].append( np.nan )
            # time diff
            data[column_name + '.time_diff.mean'].append( np.nan )
            data[column_name + '.time_diff.std'].append( np.nan )
            data[column_name + '.time_diff.min'].append( np.nan )
            data[column_name + '.time_diff.max'].append( np.nan )
            data[column_name + '.time_diff.median'].append( np.nan )
            
    # Least popular domain
    bottom = Counter( group['qname'] ).most_common()[-1]
    data['qname.least_pop.domain'].append( bottom[0] )    #domain
    data['qname.least_pop.count'].append( bottom[1] )     #occurances

    # FREQUENCY BASED FEATURES
    # TODO only for background data

In [146]:
# check num entries before converting to dataframe
for k,v in data.items():
    print len(v), "\t", k

9019 	dstIP.most_popular
9019 	qname.most_pop2.count
9019 	qname.most_pop0.time_diff.median
9019 	qname.most_pop1.count
9019 	1LD.unique.count
9019 	time_diff.median
9019 	qname.unique.count
9019 	qname.most_pop0.time_diff.mean
9019 	qname.most_pop0.count
9019 	qname.most_pop2.time_diff.median
9019 	qname.most_pop1.time_diff.min
9019 	qname.most_pop2.time_diff.min
9019 	2LD.unique.count
9019 	qname.most_pop2.time_diff.std
9019 	qname.least_pop.count
9019 	time_diff.std
9019 	qname.most_pop1.time_diff.mean
9019 	qname.most_pop1.time_diff.std
9019 	qname.most_pop0.time_diff.std
9019 	qname.most_pop2.domain
9019 	qname.most_pop0.time_diff.min
9019 	srcIP
9019 	time_diff.max
9019 	time_diff.mean
9019 	qname.most_pop1.domain
9019 	qname.most_pop0.domain
9019 	qname.count
9019 	qname.most_pop0.time_diff.max
9019 	dstIP.count
9019 	time_diff.min
9019 	qname.most_pop1.time_diff.max
9019 	qname.most_pop1.time_diff.median
9019 	qname.least_pop.domain
9019 	qname.most_pop2.time_diff.mean
9019 	qn

In [147]:
df_features = pd.DataFrame(data)
df_features.head()

1LD.unique.count  2LD.unique.count  dstIP.count dstIP.most_popular  \
0                 3                17            1     128.112.129.32   
1                 1                 1            1      128.112.128.1   
2                 1                 1            1      128.112.128.1   
3                 1                 2            1      128.112.128.1   
4                 1                 2            1      128.112.128.1   

   qname.count  qname.least_pop.count         qname.least_pop.domain  \
0           82                      1                 www.google.com   
1            6                      1  vortex-win.data.microsoft.com   
2            1                      1               dns.msftncsi.com   
3           12                      1               go.microsoft.com   
4            2                      1        ctldl.windowsupdate.com   

   qname.most_pop0.count           qname.most_pop0.domain  \
0                     12        e1863.dspb.akamaiedge.net   
1                      5  settings-win.data.microsoft.com   
2                      1                 dns.msftncsi.com   
3                      6    statsfe2.update.microsoft.com   
4                      1                crl.microsoft.com   

   qname.most_pop0.time_diff.max      ...        \
0                     826.509368      ...         
1                     964.020238      ...         
2                            NaN      ...         
3                     541.406385      ...         
4                            NaN      ...         

   qname.most_pop2.time_diff.median  qname.most_pop2.time_diff.min  \
0                        225.109573                       5.840552   
1                               NaN                            NaN   
2                               NaN                            NaN   
3                               NaN                            NaN   
4                               NaN                            NaN   

   qname.most_pop2.time_diff.std  qname.unique.count        srcIP  \
0                     458.755275                  23   10.6.48.10   
1                            NaN                   2  10.6.48.100   
2                            NaN                   1  10.6.48.105   
3                            NaN                   3  10.6.48.112   
4                            NaN                   2  10.6.48.116   

  time_diff.max  time_diff.mean  time_diff.median  time_diff.min  \
0    226.720494       42.408970         27.007246       0.017376   
1    964.020238      584.807132        963.999243       0.000284   
2           NaN             NaN               NaN            NaN   
3    541.365839      262.413738        180.582701      76.617629   
4      0.009410        0.009410          0.009410       0.009410   

   time_diff.std  
0      50.710004  
1     519.369585  
2            NaN  
3     193.950951  
4            NaN  

[5 rows x 35 columns]

## 3. FILTER devices with <=10 unique 2LDs

In [148]:
df_filtered1 = df_features[(df_features['2LD.unique.count']<=10)]

print len(df_filtered1)

4037


## 4. FILTER devices with >10 total queries throughout the 1 hour pcap

In [149]:
df_filtered2 = df_filtered1[(df_filtered1['qname.count']>10)]

print len(df_filtered2)

2223


## Set key as srcIP and join with srcIP_groundtruth table

In [153]:
df_full = df_filtered2.reset_index().set_index('srcIP').join(srcIP_groundtruth)
del df_full['index']

In [161]:
df_full.head()

1LD.unique.count  2LD.unique.count  dstIP.count  \
srcIP                                                          
10.6.48.112                 1                 2            1   
10.6.48.13                  2                 7            2   
10.6.48.14                  1                 1            1   
10.6.48.148                 2                 4            2   
10.6.48.160                 1                 1            2   

            dstIP.most_popular  qname.count  qname.least_pop.count  \
srcIP                                                                
10.6.48.112      128.112.128.1           12                      1   
10.6.48.13       128.112.128.1          105                      1   
10.6.48.14       128.112.128.1          131                    131   
10.6.48.148      128.112.128.1           24                      3   
10.6.48.160      128.112.129.7           12                     12   

                  qname.least_pop.domain  qname.most_pop0.count  \
srcIP                                                             
10.6.48.112             go.microsoft.com                      6   
10.6.48.13              go.microsoft.com                     28   
10.6.48.14   endpoint.ingress.rapid7.com                    131   
10.6.48.148             dns.msftncsi.com                      6   
10.6.48.160      linux-update.oracle.com                     12   

                    qname.most_pop0.domain  qname.most_pop0.time_diff.max  \
srcIP                                                                       
10.6.48.112  statsfe2.update.microsoft.com                     541.406385   
10.6.48.13             clients1.google.com                     180.060638   
10.6.48.14     endpoint.ingress.rapid7.com                      40.077696   
10.6.48.148  statsfe2.update.microsoft.com                     601.741539   
10.6.48.160        linux-update.oracle.com                      45.112617   

                   ...         time_diff.median  time_diff.min  time_diff.std  \
srcIP              ...                                                          
10.6.48.112        ...               180.582701      76.617629     193.950951   
10.6.48.13         ...                29.255084       0.003357      35.207985   
10.6.48.14         ...                20.093167      10.030509      12.012770   
10.6.48.148        ...               131.192425       0.024489     131.158580   
10.6.48.160        ...                15.001348       0.000710      20.221434   

                      macAddr          subnet host_type       host_name  \
srcIP                                                                     
10.6.48.112   0:50:56:93:6a:6  fw-servernet-2      host      spcacs201w   
10.6.48.13   0:50:56:93:5e:54  fw-servernet-2      host  essconsole200w   
10.6.48.14   0:50:56:93:5e:45  fw-servernet-2      host  ess-2008r2test   
10.6.48.148  0:50:56:93:17:5f  fw-servernet-2      host          ss220w   
10.6.48.160  0:50:56:93:5d:c3  fw-servernet-2      host         ais104l   

                 system_type               OS  nic_manufacturer  
srcIP                                                            
10.6.48.112  virtual-machine  ms-windows-2012            Vmware  
10.6.48.13   virtual-machine  ms-windows-2012            Vmware  
10.6.48.14   virtual-machine  ms-windows-2008            Vmware  
10.6.48.148  virtual-machine  ms-windows-2012            Vmware  
10.6.48.160  virtual-machine            linux            Vmware  

[5 rows x 41 columns]

## 5. FILTER based on host_type and system_type
- system_type != virtual-machine

In [158]:
vms = df_full[df_full['system_type']=='virtual-machine']
print len(vms)

122


In [159]:
df_hosts = df_full[df_full['system_type']!='virtual-machine']

In [163]:
df_wireless = df_hosts[df_hosts['host_type']=='host-dynamic']
df_wired = df_hosts[df_hosts['host_type']!='host-dynamic']

print len(df_wired), len(df_wireless)

9 2092


In [164]:
df_wired

1LD.unique.count  2LD.unique.count  dstIP.count  \
srcIP                                                             
10.6.48.62                     1                 1            2   
10.6.49.1                      2                 8            1   
10.6.49.172                    1                 2            1   
10.6.52.121                    1                 1            2   
10.6.52.125                    3                 3            2   
10.6.63.102                    1                 1            2   
140.180.219.15                 1                 1            1   
172.20.50.101                  2                 2            1   
172.20.50.102                  2                 2            1   

               dstIP.most_popular  qname.count  qname.least_pop.count  \
srcIP                                                                   
10.6.48.62          128.112.129.7           12                     12   
10.6.49.1           128.112.128.1           50                      1   
10.6.49.172       128.112.129.111          175                     12   
10.6.52.121         128.112.129.7           12                     12   
10.6.52.125         128.112.129.7           31                      1   
10.6.63.102         128.112.129.7           24                     24   
140.180.219.15    128.112.133.220          147                    147   
172.20.50.101         10.6.48.144          214                      8   
172.20.50.102         10.6.48.144          214                      8   

                 qname.least_pop.domain  qname.most_pop0.count  \
srcIP                                                            
10.6.48.62      linux-update.oracle.com                     12   
10.6.49.1           clients2.google.com                     10   
10.6.49.172     g4t2482g.houston.hp.com                     79   
10.6.52.121      feeds.travel.state.gov                     12   
10.6.52.125                ftp.ecsi.net                     18   
10.6.63.102     linux-update.oracle.com                     24   
140.180.219.15    outlook.office365.com                    147   
172.20.50.101     3.debian.pool.ntp.org                    182   
172.20.50.102     3.debian.pool.ntp.org                    182   

                          qname.most_pop0.domain  \
srcIP                                              
10.6.48.62               linux-update.oracle.com   
10.6.49.1         clients2.googleusercontent.com   
10.6.49.172     storage-support.glb.itcs.hpe.com   
10.6.52.121               feeds.travel.state.gov   
10.6.52.125              www.concursolutions.com   
10.6.63.102              linux-update.oracle.com   
140.180.219.15             outlook.office365.com   
172.20.50.101                 smtp.office365.com   
172.20.50.102                 smtp.office365.com   

                qname.most_pop0.time_diff.max        ...         \
srcIP                                                ...          
10.6.48.62                          18.000263        ...          
10.6.49.1                          584.209087        ...          
10.6.49.172                         60.008996        ...          
10.6.52.121                        160.021907        ...          
10.6.52.125                        315.553191        ...          
10.6.63.102                         45.082981        ...          
140.180.219.15                     305.171067        ...          
172.20.50.101                      120.728163        ...          
172.20.50.102                      120.619818        ...          

                time_diff.median  time_diff.min  time_diff.std  \
srcIP                                                            
10.6.48.62             15.001484       0.001044       7.236358   
10.6.49.1              39.639565       0.002616      80.158844   
10.6.49.172             0.095821       0.003952      28.022075   
10.6.52.121             0.002988       0.000618      63.535062   
10.6.52.125            74.900722       0

In [166]:
len(df_wireless)

2092

In [167]:
df_wireless.subnet.unique()

array(['vapornet100'], dtype=object)

In [174]:
len( df_wireless[ df_wireless['macAddr'].isnull()] )

1443

In [168]:
df_wireless.nic_manufacturer.unique()

array([nan, 'Apple', 'GioneeCo', 'LiteonTe', 'IntelCor', 'SamsungE',
       'LgElectr', 'Microsof', 'MurataMa', 'XiaomiCo', 'Fn-LinkT',
       'Motorola', 'AmazonTe', 'HonHaiPr', 'HuaweiTe', 'TctMobil', 'Htc',
       'Barnes&N', 'Azurewav', 'ChiconyE', 'Private', 'BluProdu',
       'OneplusT', 'SonyMobi', 'Zte'], dtype=object)

## Summary of hosts
- df_wired contains 9 unique IPs.
    - 7 of these seem to be departmental computers on servernet subnet running linux, windows or VMware.
    - 2 IPs (172.20.50.101, 172.20.50.102) seem to be running raspberry pi systems in the engineering building performing the same number of queries to ntp.org and office365.com
- df_wireless contains 2092 unique IPs.
    - all of these are NATted over the vapornet100 subnet
    - 1443 of 2092 IPs do not have a corresponding DHCP entry containing MAC address

## Analyzing wireless hosts
- TODO: find periods for all domain queries for these hosts

In [179]:
df_wireless[['macAddr', 'nic_manufacturer', 'qname.count', 'qname.unique.count', '1LD.unique.count',
             '2LD.unique.count', 'qname.most_pop0.domain', 'qname.most_pop0.time_diff.mean', 'qname.most_pop0.time_diff.std' ]]

macAddr nic_manufacturer  qname.count  \
srcIP                                                            
10.8.0.150                   NaN              NaN           90   
10.8.0.213                   NaN              NaN           28   
10.8.0.39                    NaN              NaN           21   
10.8.1.70                    NaN              NaN           29   
10.8.1.83                    NaN              NaN           11   
10.8.1.86      0c:15:39:31:ef:67            Apple           42   
10.8.10.12                   NaN              NaN           31   
10.8.10.165    18:65:90:3b:d2:e2            Apple           37   
10.8.10.226                  NaN              NaN           22   
10.8.10.56     2c:f0:ee:13:cd:58            Apple           22   
10.8.10.69                   NaN              NaN           35   
10.8.100.19    60:f8:1d:bc:bd:a8            Apple           31   
10.8.100.22   6c:8d:c1:d5:100:1d            Apple           35   
10.8.101.141                 NaN              NaN           42   
10.8.101.200                 NaN              NaN           17   
10.8.101.237                 NaN              NaN          153   
10.8.101.98    4c:7c:5f:a2:d0:70            Apple           33   
10.8.102.146                 NaN              NaN           12   
10.8.102.17                  NaN              NaN           20   
10.8.102.194                 NaN              NaN           16   
10.8.102.235   9c:f3:87:bc:6c:6a            Apple           12   
10.8.102.44                  NaN              NaN           46   
10.8.102.82                  NaN              NaN           25   
10.8.103.132    64:b0:a6:a:c2:21            Apple           57   
10.8.103.141                 NaN              NaN           32   
10.8.103.45                  NaN              NaN           13   
10.8.104.174                 NaN              NaN           38   
10.8.104.200   54:9f:13:f2:2b:6f            Apple           22   
10.8.104.208                 NaN              NaN           35   
10.8.104.22                  NaN              NaN           12   
...                          ...              ...          ...   
10.9.93.232    d0:03:4b:c0:2a:2d            Apple          156   
10.9.94.101    1c:91:48:73:80:33            Apple           24   
10.9.94.187    f4:0f:24:30:88:28            Apple           11   
10.9.94.231                  NaN              NaN           11   
10.9.94.36                   NaN              NaN           52   
10.9.95.122                  NaN              NaN          116   
10.9.95.123    b8:e8:56:4b:ca:5a            Apple           38   
10.9.95.22     d8:cf:9c:e9:78:80            Apple           18   
10.9.95.252                  NaN              NaN           30   
10.9.96.14     48:3b:38:24:c8:b4            Apple           24   
10.9.96.37                   NaN              NaN           29   
10.9.97.117    00:b3:62:4e:bf:ce            Apple           49   
10.9.97.12                   NaN              NaN           13   
10.9.97.218                  NaN              NaN           43   
10.9.97.253                  NaN              NaN           14   
10.9.97.35                   NaN              NaN           14   
10.9.97.76                   NaN              NaN           18   
10.9.98.125                  NaN              NaN           26   
10.9.98.184    fc:c2:de:8c:ba:9f         MurataMa           31   
10.9.98.242                  NaN              NaN           21   
10.9.98.81                   NaN              NaN           12   
10.9.99.112                  NaN              NaN           13   
10.9.99.133    a4:5e:60:ca:ae:61            Apple           11   
10.9.99.134    78:4f:43:72:c1:21            Apple           11   
10.9.99.137    a8:88:08:aa:55:13            Apple           31   
10.9.99.223    b0:34:95:8d:f1:97            Apple           37   
10.9.99.232                  NaN              NaN           44   
10.9.99.237    98:01:a7:9c:bc:53            Apple           12   

# CLUSTERING